### PR TITLE
refactor(npm-db): update author field to be more verbose

### PR DESCRIPTION
**Pull request by pxlpnk**
This splits up the author field into 3 fields
- name
- website
- username

``` js
"author": {
    "name": "Cal Leeming", // required
    "website": null, // optional, can be null
    "username": null //  optional, can be null
}
```

The PR comes with a (`tools/migrations/transform_author.js`) script that changes the existing database.
You can check out the branch and run: `node tools/migrations/transform_author.js`). Then run `npm test` afterwards to check the migration.

The migrated files where not yet checked in to keep the PR  less noisy. I will supply those changes as soon as the proposed changes are approved. The PR will be failing until those changes are applied.

I updated the files with id 109 and 113 as they had author set to null and replaced it with `unknown`.

This is the first PR to get #200 moving forward.

### CODE REVIEW

File: tools/vuln_valid/index.js [hunk 1]
```diff
@@ -24,7 +24,14 @@ const npmModel = joi.object().keys({
   updated_at: joi.string().regex(/^\d{4}-\d{2}-\d{2}$/).required().isoDate(),
   title: joi.string().required(),
   title: joi.string().max(150).regex(/^[^\n]+$/).required(),
-  author: joi.string().allow(null).required(),
+  author: joi.object().required(),
+  author: joi.object().keys(
+    {
+      name: joi.string().required(),
+      username: joi.string().optional().allow(null),
```
Comment (pxlpnk) [inline tools/vuln_valid/index.js]: Should username and website be required, but allowed to be null?
I kind of think this makes more sense.

Comment (lirantal) [reply tools/vuln_valid/index.js]: It helps consistency through the reports format so that's ok with me.

**Comment by pxlpnk** [comment]
Rebased and migrated the files

File: vuln/npm/129.json [hunk 2]
```diff
@@ -17,4 +21,4 @@
   "cvss_vector": "CVSS:3.0/AV:N/AC:L/PR:L/UI:R/S:U/C:H/I:H/A:N",
   "cvss_score": 7.3,
   "coordinating_vendor": "^Lift Security"
-}
+}
```
Comment (vdeturckheim) [inline vuln/npm/129.json]: Let's keep the empty line

File: vuln/npm_old/98.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  "overview": "The primary npm registry has, since late 2014, used HTTP bearer tokens to authenticate requests from the npm command-line interface. Due to a design flaw in the CLI, these bearer tokens were sent with every request made by the CLI for logged-in users, regardless of the destination of the request. They should instead only be included for requests made against the registry or registries used for the current install.\n\nThis flaw allows an attacker to set up an HTTP server that could collect authentication information they could use to impersonate the users whose tokens they collected. This impersonation would allow them to do anything the compromised users could do, including publishing new versions of packages.",
+  "recommendation": "Upgrade npm itself with `npm install npm@latest -g` and [revoke your tokens](https://www.npmjs.com/settings/tokens)",
+  "references": "- https://nodejs.org/en/blog/vulnerability/npm-tokens-leak-march-2016/\n- http://blog.npmjs.org/post/142036323955/fixing-a-bearer-token-vulnerability",
+  "cvss_vector": "CVSS:3.0/AV:N/AC:H/PR:N/UI:R/S:U/C:H/I:H/A:N",
+  "cvss_score": 6.8,
+  "coordinating_vendor": "^Lift Security"
+}
```
Comment (vdeturckheim) [inline vuln/npm_old/98.json]: What should we do with this directory?

File: vuln/npm/129.json (hunk 2 above)
Comment (pxlpnk) [reply vuln/npm/129.json]: 👍 Updated the PR

File: vuln/npm_old/98.json (hunk 3 above)
Comment (pxlpnk) [reply vuln/npm_old/98.json]: /me fat-fingered here.

**Comment by pxlpnk** [comment]
I think this one and #316 should be in good shape. WDYTH?

**Comment by lirantal** [comment]
@pxlpnk sounds good but let me also prepare a PR to our automatic report creation script at https://github.com/nodejs/security-wg/pull/234 to accommodate for the changes.

I do also want to get feedback about both of these in the main issue before we merge.

File: tools/vuln_valid/index.js [hunk 4]
```diff
@@ -24,7 +24,14 @@ const npmModel = joi.object().keys({
   updated_at: joi.string().regex(/^\d{4}-\d{2}-\d{2}$/).required().isoDate(),
   title: joi.string().required(),
   title: joi.string().max(150).regex(/^[^\n]+$/).required(),
-  author: joi.string().allow(null).required(),
+  author: joi.object().required(),
```
Comment (lirantal) [inline tools/vuln_valid/index.js]: the first author field is redundant, no?
same goes for the title field above

Comment (pxlpnk) [reply tools/vuln_valid/index.js]: You where right, I removed the first `author: joi.object().required()`. As for the `title` field. I don't know what the intention with this was. But I can clean this up in a separate PR. I think there might be more in this file that could need a touch up.

Comment (lirantal) [reply tools/vuln_valid/index.js]: Sure, np

**Pull request merged**